apps/aecore/src/aec_hard_forks.erl

```diff
@@ -98,24 +115,102 @@ protocols_from_network_id(_ID) ->
+    undefined.
+    %% #{signalling_start_height => undefined,
+    %%   signalling_end_height   => undefined,
+    %%   signalling_block_count  => undefined,
+    %%   info_field              => undefined,
+    %%   version                 => undefined}.
+
+testnet_fork_config() ->
+    undefined.
+    %% #{signalling_start_height => undefined,
+    %%   signalling_end_height   => undefined,
+    %%   signalling_block_count  => undefined,
+    %%   info_field              => undefined,
+    %%   version                 => undefined}.
+
 %% Exported for tests
 sorted_protocol_versions() ->
     lists:sort(maps:keys(protocols())).
 
 protocols_sorted_by_version(Ps) ->
     lists:keysort(1, maps:to_list(Ps)).
 
-protocol_effective_at_height(H, Protocols) ->
-    assert_height(H),
+protocol_effective_at_height(Height, Protocols) ->
+    assert_height(Height),
     SortedProtocols = protocols_sorted_by_version(Protocols),
     %% Find the last protocol version effective before or at the
     %% specified height: that is the one effective at the specified
     %% height.  This assumes that the height is strictly increasing
     %% with the version so also assert that for the sake of clarity.
     ProtocolsEffectiveSinceBeforeOrAtHeight = [_|_] =
-        lists:takewhile(fun({_, HH}) -> H >= HH end, SortedProtocols),
-    {V, _} = lists:last(ProtocolsEffectiveSinceBeforeOrAtHeight),
-    V.
+        lists:takewhile(fun({_, H}) -> Height >= H end, SortedProtocols),
+    {Protocol, _ForkHeight} = lists:last(ProtocolsEffectiveSinceBeforeOrAtHeight),
+    maybe_protocol_from_fork(aeu_env:get_env(aecore, fork), Height, Protocol).
+
+maybe_protocol_from_fork(undefined, _Heigth, Protocol) ->
+    %% No community fork configured.
+    Protocol;
+maybe_protocol_from_fork({ok, #{signalling_end_height := EndHeight}}, Height, Protocol)
+  when Height < EndHeight ->
+    %% Height is lower than signalling end height, so the greatest protocol
+    %% before the community fork is returned.
+    Protocol;
+maybe_protocol_from_fork({ok, #{signalling_end_height := Height} = Fork}, Height, Protocol) ->
+    %% Height is equal to the signalling end height which is the fork heght. The
+    %% fork signalling result needs to be taken from aec_chain_state module for
+    %% the last key block before the fork height (last signalling block height).
+    {ok, Block} = aec_chain:get_key_block_by_height(Height - 1),
+    case aec_chain_state:get_fork_result(Block, Fork) of
+        {ok, true}  -> maps:get(version, Fork);
+        {ok, false} -> Protocol
+    end;
+maybe_protocol_from_fork({ok, #{signalling_end_height := EndHeight}}, Height, _Protocol)
+  when Height > EndHeight ->
+    %% Height is greater than the signalling end height. The decision about the
+    %% protocol version already happened, so the protocol version at this height
+    %% is the same as the protocol version of the previous (key) block.
+    {ok, Block} = aec_chain:get_key_block_by_height(Height - 1),
```

Comment:
Hmm, looking up a block from a height is slightly expensive... 

I did check, and in all places where we do (now) call `protocol_effective_at_height` we already have the current block, and thus easy access to its `prev_blockhash`, I wonder if it would make sense to pass this information and save some computation here :thinking:

Reply:
Getting a block by height is called only for a block at fork height. I think I'd have to change the API of `protocol_effective_at_height` and instead of `Height` pass `Block` there - not sure it's worth it or possible? `protocol_effective_at_height` is called also when key block candidate is prepared and in that case there is no block to pass yet, just height.

Reply:
Not `block` itself, but previous blockhash?!
And...
```
create(Block, Beneficiary) ->
    {ok, BlockHash} = aec_blocks:hash_internal_representation(Block),
    Height = aec_blocks:height(Block) + 1,
    Protocol = aec_hard_forks:protocol_effective_at_height(Height),
    int_create(BlockHash, Block, Beneficiary, Protocol).
```

Reply:
yes, that looks doable, but also I need `Height` (to get protocol for scheduled forks):
```
protocol_effective_at_height(Height, Protocols) ->
    assert_height(Height),
    SortedProtocols = protocols_sorted_by_version(Protocols),
    %% Find the last protocol version effective before or at the
    %% specified height: that is the one effective at the specified
    %% height.  This assumes that the height is strictly increasing
    %% with the version so also assert that for the sake of clarity.
    ProtocolsEffectiveSinceBeforeOrAtHeight = [_|_] =
        lists:takewhile(fun({_, H}) -> Height >= H end, SortedProtocols),
    {Protocol, _ForkHeight} = lists:last(ProtocolsEffectiveSinceBeforeOrAtHeight),
    maybe_protocol_from_fork(aeu_env:get_env(aecore, fork), Height, Protocol).
```

Reply:
Yes, `(Height, PrevBlockHash)` would be my suggestion... But this isn't crucial, just started thinking when I saw the code.